### ISSUE TITLE
Approach to applying flake8 / isort / black (or blue) on all files in the package (and what tools should we use)

### ISSUE THREAD

**Issue by lwasser** [stravalib/stravalib#293]
Ok I think my current documentation PR is really close to being merged. So i've started to play with code formatters and stylers locally. 

This is going to potentially create ugly merge conflicts so it would be better if i work on files that others are NOT working on one or two at a time. To ensure things go smoothly.

- [ ] _version.py
- [ ] attributes.py
- [ ] client.py
- [ ] exc.py
- [ ] model.py
- [ ] protocol.py
- [ ] unithelper.py

@jsamoocha @yihong0618 which of the above files would be best for me to start working on? i can do one or two at a time submit a PR and then they'd be good to have more work done on them. We can keep in touch on this change here to ensure no one is working on files i'm formatting locally. and i'll do my best to implement the format change in 1-2 days to not hold up work. Which means i won't start a new file until i know i have time to fix it!

i'll:

* run black (or blue) on them and 
* make sure they pass flake8 tests 
*  implement isort (i'm less worried about isort as it's just imports)

in ONE pr i'll also add an precommit hook that is optional to setup just in case we want to format on commits locally (vs running make)

**Comment by lwasser** [comment]
related to this. Who has opinions on code formatters here?! 

1. [ruff](https://github.com/charliermarsh/ruff) is a really cool new tool that is written in rust. it's speedy fast but doesn't implement a few flake8 things. but it also runs isort checks and a bunch of other checks. 
2. [flake8 ](https://flake8.pycqa.org/en/latest/)is another OG but it only lets us know what is wrong. we also have to configure it in the setup.cfg file rather than the pyproject.toml file which is a bit annoying but ok. 
3. [black is a code styler](https://black.readthedocs.io/en/stable/) - it's the OG of code stylers and i'm a fan. But some don't like those darn double quotes that it puts everywhere.
4. [blue](https://blue.readthedocs.io/en/latest/) addresses the things that annoy some people about black. it also has a pretty amazing icon image on the home page of the documentation 😆 
***
from the docs here is what blue says differs from black

```
So what’s different?
Here is a brief list of differences between blue and black:

blue defaults to single-quoted strings. This is probably the most painful black choice to our eyes, and the thing that inspired blue. We strongly prefer using single quoted strings over double quoted strings for everything except docstrings and triple quoted strings (TQS). Don’t ask us why we prefer double-quotes for TQS; it just looks better to us! For all other strings, blue defaults to single quoted strings. In the case of docstrings, those always use TQS with double-quotes.
blue defaults to line lengths of 79 characters. Nearly every project creates a pyproject.toml just to change this one setting so making it consistent with PEP 8 seems relatively harmless.
blue preserves the whitespace before the hash mark for right hanging comments.
blue supports multiple config files: pyproject.toml, setup.cfg, tox.ini, and .blue.
```

i'm leaning towards starting with 

* blue
* flake8
* isort
and seeing how it goes given ruff as cool as it is isn't quite fully operational yet. And blue has some of the configurations that i make in black such as line length 79 (following pep 8).

ut i'm open to other folks thoughts. i just don't want to impose my opinions on this if others have stronger conflicting opinions. many thanks for thoughts!

**Comment by yihong0618** [comment]
I think `black` with `isort` is the easiest one to use.

**Comment by jsamoocha** [comment]
I don't think I have a preference for a specific formatter. Also, I'm not planning to work on any PR this week, so the risk of nasty merge conflicts will be minimal.

For my understanding: if the formatting pre-commit hook is optional and not applied, will the build action create a new (formatting) commit when a PR is submitted? Or will the build fail and ask the PR author to format the code?

**Comment by lwasser** [comment]
hey everyone! so sorry i've had to focus on other things this week. But i hope to work on format / linting the code this week. 

So @yihong0618 one note... but i'm also happy to use black. [Blue is actually a wrapper for black. click here for docs](https://blue.readthedocs.io/en/latest/#:~:text=What%20is%20blue%3F,of%20the%20decisions%20black%20makes.)

It's nice because it changes a few of the annoying things about black as follows (but otherwise is black with just a few modifications - the docs even link to black docs!!)

```
blue defaults to single-quoted strings. This is probably the most painful black choice to our eyes, and the thing that inspired blue. We strongly prefer using single quoted strings over double quoted strings for everything except docstrings and triple quoted strings (TQS). Don’t ask us why we prefer double-quotes for TQS; it just looks better to us! For all other strings, blue defaults to single quoted strings. In the case of docstrings, those always use TQS with double-quotes.
* blue defaults to line lengths of 79 characters. Nearly every project creates a pyproject.toml just to change this one setting so making it consistent with PEP 8 seems relatively harmless.
* blue preserves the whitespace before the hash mark for right hanging comments.
* blue supports multiple config files: pyproject.toml, setup.cfg, tox.ini, and .blue.
```

so essentially i don't have to set 79 characters (which is a pep 8 requirement because blue already defaults to that (so it's just one less thing to setup. But under the hood it's black! Lemme know what you think about that. 

In the meantime i'll start working on implementing flake8, black (or blue if you are ok with that!??) and isort which is for me to play with!!

I'll submit pr's that are just a 1-2 files at a time so they will be small. but will add only one line to the changelog! Please post here if you plan to work on a file this week i'll avoid messing with it to avoid merge conflicts!! i should be able to do a file or two a day this week!

Thank you both for the feedback here!

****
Regarding the pre-commit hook @jsamoocha this hook is one that happens locally on your computer as you commit to git. So you run git commit, whatever hooks you have setup (i will add instructions) will run. This would normally be black/blue and flake8. you won't be able to actually commit until you fix format issues. 

this just helps enforce that your code is styled correctly before pushing to CI at all. 

But it's totally optional. you might chose to not install that hook. you could have black or blue running in pycharm for example each time you save. so it's totally optional. Let me know what you think about having it. i wouldn't mind it for myself but again you don't have to install it - i can just provide instructions for how it works for those who like that workflow.

Please let me know if you have any other questions, concerns etc!!

**Comment by lwasser** [comment]
Ok based on all of the conversations I will submit a PR that does a few small / big things.

1. i'll run black on the entire code base (@jsamoocha is that going to mess with your work on pydantic - if so i won't run it)
2. I'll setup a pre-commit hook and update docs. the hook will be for flake8, black and isort... flake8 is the one that is going to be a PITA as there are literally hundreds of violations in the code now. so we COULD hold off on applying that hook until the big model change going through.  (also the hook is totally optional in your local dev workflow)

let me know if that sounds reasonable. And i guess the only question is - should i hold off actually running black on the entire code base? it's easy and quick but will change literally every line of code in the repo. 

Happy early new year!

**Comment by lwasser** [comment]
@jsamoocha i'm hesitant to have the build actually run black -- i'll talk to some colleagues about that but normally we just run the pre-commit, check things / fix manually and then push changes that way. i'll try to prioritize this specific PR this week as it will be fairly small especially if i don't actually lint / cleanup the files.

**Comment by jsamoocha** [comment]
Sounds good. Just be aware that as a result of the pydantic model work, the file `attributes.py` will be gone, and the file `model.py` will have major changes.

**Comment by lwasser** [comment]
ok @jsamoocha  [please have a look at this PR open in my fork.](https://github.com/lwasser/stravalib/pull/19/commits)

I'm setting pre-commit.ci up. [I am told by Jupyter devs that it works really well ](https://github.com/jupyterhub/team-compass/issues/379).<-  see convo about the tool there. 

I set this up to just run isort on the repo. And you can see how the bot DOES update my pr. 

DANGER ZONE: we would want to trial this in case it could cause merge conflicts. most of the time - once we have this all run on all files in the repo, the commit edits would be tiny. in this case these tools have never been run so it's making a lot of changes.  I suspect it could be great for us. 

we can also turn off the auto pr's or modify how that bot runs!
* i have NOT run black yet - just isort

Should we give this a try? 
and also i understand model.py will be modified significantly and attributes.py will be gone. 

Should i run black on all other files now? 
(so i'd run it on everything BUT model.py).

Then you can run on model.py locally as you work? it's fast to run and would update the code as you work.

Let me know what you think about both the PR in my repo and this approach in general. it's a nice way to avoid forcing new contributors to deal with pre-commit stuff.  but we will have to see if it creates pr merge conflict messes.

**Comment by jsamoocha** [comment]
> Should we give this a try?

Why not? It looks useful.

While we already have a "test" branch (i.e. `pydantic_model`), we could merge your PR into that branch and see how it behaves on subsequent changes/PRs I will be adding to it later. That way we can keep the main branch stable. What do you think?

**Comment by lwasser** [comment]
ok great! i haven't added black to it. should i hold off on that and just submit as is to pydantic_model  branch? I'm asking because i know that black will really modify all of our modules. and i don't want you to deal with a massive conflict. BUT you could actually run black locally first. I can run it on modules that you aren't touching now to minimize the effect of when we add the hook.

Does that make sense?

I'm trying to avoid a PR mess to clean on your end here :)

i think in the long run this bot will be great. i need to setup it up here next but you can call it in any pr to run manually too. it's pretty cool. then no new comer contributors need to understand how to setup a pre-commit hook locally! yay! (but we can use it ourselves in our workflows)

**Comment by jsamoocha** [comment]
I haven't touched any code in `pydantic_model` since my last PR, so there will be no merge conflicts. Feel free to add black and let it modify everything; it needs to happen sooner or later.

If there are important changes to `master` before we finish work on `pydantic_model`, we can always cherry-pick and merge these changes into `pydantic_model`. Wouldn't that prevent nasty merge conflicts from happening (but maybe I'm missing something)?

**Comment by lwasser** [comment]
ok i've merged #306 into the `pydantic_model` branch! i'll leave this open as i forgot to update the changelog with that merge but also it didn't go to main, yet!  we can close this once it's in main and added to the changelog! so far it's really nice to have the bot doing our work!
